redis/cache/redisx/redisxtest/redisxtest.go

```diff
@@ -0,0 +1,70 @@
+package redisxtest
+
+import (
+	"context"
+
+	"github.com/alicebob/miniredis/v2"
+	"github.com/joomcode/redispipe/redis"
+	"github.com/joomcode/redispipe/redisconn"
+
+	"github.com/reddit/baseplate.go/redis/cache/redisx"
+)
+
+// MockRedisCluster wraps a local version of redis
+type MockRedisCluster struct {
+	redisCluster *miniredis.Miniredis
+}
+
+func NewMockRedisCluster() (mockRedisCluster MockRedisCluster, teardown func(), err error) {
+	redisCluster, err := miniredis.Run()
+	if err != nil {
+		return MockRedisCluster{}, nil, err
+	}
+
+	mockRedisCluster = MockRedisCluster{
+		redisCluster: redisCluster,
+	}
+
+	teardown = func() {
+		mockRedisCluster.Close()
+	}
+
+	return mockRedisCluster, teardown, nil
+}
+
+// Addr returns address of mock redis cluster e.g. '127.0.0.1:12345'.
+func (mrc *MockRedisCluster) Addr() string {
+	return mrc.redisCluster.Addr()
+}
+
+// Close shuts down the MockRedisCluster
+func (mrc *MockRedisCluster) Close() {
+	mrc.redisCluster.Close()
+}
+
+// NewMockRedisClient sets up a client and sender to a mock redis cluster
+func NewMockRedisClient(
+	ctx context.Context,
+	redisCluster MockRedisCluster,
```

Comment:
this feels like an over-design to me.

if in most tests we only need a single, global, mocked cluster and client, then we don't really need separated `NewMockRedisCluster` and `NewMockRedisClient`, a single `NewMockRedisClient` with its returned `teardown` also tears down the cluster should be sufficient (the cluster does not need to be exposed to the caller at all).

but if that's not the case, then I think only the cluster itself should be global (that's handled inside `TestMain`), the clients should be created per test, so my previous comment regarding passing in `testing.TB` and no need to return `error` nor `teardown` applies.

Reply:
and looking at this function again I don't see it doing anything special. all it does are pretty standard stuff about create a redis connection (you actually only need the address passed in, not the whole cluster object) and create a redis client out of this connection. the `teardown` is also standard stuff about closing the client, so I don't really see much value of providing (and maintaining) this as part of the public API.

If we think a `func NewClient(ctx, addr, opts) (redisx.BaseSync, error)` API is useful, then that should be part of `redisx`, not `redisxtest`.

Reply:
Reason for the change to have separated `NewMockRedisCluster` and `NewMockRedisClient` was to support the cases like that in `TestMain` in `init_test`, where it looks like we need a global cluster. Perhaps an alternative solution to this would be to have the following API for redisxtest:

```
func NewMockRedisClient(tb, ctx) (client, error)
func NewMockRedisClientWithCluster(tb, ctx, clusterAddr) (client, error)
```
These functions would handle cleanup and client creation via tb as you have suggested. 

I am really not opposed at all to going down the path updating the redisx API to have a `func NewClient()`. This is actually our preferred route. We saw adding `redisxtest` as being minimally invasive, which was the reason for going this path. If you are ok with us doing that, happy to open another PR for that change and close potentially close this one if we see no value in what I have proposed.